stores/settingsdb.go

```diff
@@ -2,39 +2,133 @@ package stores
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 
+	"go.sia.tech/renterd/api"
 	sql "go.sia.tech/renterd/stores/sql"
 )
 
-// DeleteSetting implements the bus.SettingStore interface.
-func (s *SQLStore) DeleteSetting(ctx context.Context, key string) error {
-	s.settingsMu.Lock()
-	defer s.settingsMu.Unlock()
+const (
+	SettingGouging = "gouging"
+	SettingPinned  = "pinned"
+	SettingS3      = "s3"
+	SettingUpload  = "upload"
+)
+
+func (s *SQLStore) GougingSettings(ctx context.Context) (gs api.GougingSettings, _ error) {
+	value, err := s.fetchSetting(ctx, SettingGouging)
+	if err != nil {
+		return api.GougingSettings{}, err
+	}
+
+	if err := json.Unmarshal([]byte(value), &gs); err != nil {
+		s.logger.Panicf("failed to unmarshal gouging settings '%s': %v", value, err)
+		return api.GougingSettings{}, err
+	}
+	return
+}
+
+func (s *SQLStore) UpdateGougingSettings(ctx context.Context, gs api.GougingSettings) error {
+	data, err := json.Marshal(gs)
+	if err != nil {
+		return fmt.Errorf("couldn't marshal the given value, error: %v", err)
+	}
+	return s.updateSetting(ctx, SettingGouging, string(data))
+}
 
-	// delete from database first
-	if err := s.db.Transaction(ctx, func(tx sql.DatabaseTx) error {
-		return tx.DeleteSettings(ctx, key)
-	}); err != nil {
+func (s *SQLStore) PinnedSettings(ctx context.Context) (ps api.PinnedSettings, _ error) {
+	value, err := s.fetchSetting(ctx, SettingPinned)
+	if err != nil {
+		return api.PinnedSettings{}, err
+	}
+
+	if err := json.Unmarshal([]byte(value), &ps); err != nil {
+		s.logger.Panicf("failed to unmarshal pinned settings '%s': %v", value, err)
+		return api.PinnedSettings{}, err
+	}
+	return
+}
+
+func (s *SQLStore) UpdatePinnedSettings(ctx context.Context, ps api.PinnedSettings) error {
+	data, err := json.Marshal(ps)
+	if err != nil {
+		return fmt.Errorf("couldn't marshal the given value, error: %v", err)
+	}
+	return s.updateSetting(ctx, SettingPinned, string(data))
+}
+
+func (s *SQLStore) UploadSettings(ctx context.Context) (us api.UploadSettings, _ error) {
+	value, err := s.fetchSetting(ctx, SettingUpload)
+	if err != nil {
+		return api.UploadSettings{}, err
+	}
+
+	if err := json.Unmarshal([]byte(value), &us); err != nil {
+		s.logger.Panicf("failed to unmarshal upload settings '%s': %v", value, err)
+		return api.UploadSettings{}, err
+	}
+	return
+}
+
+func (s *SQLStore) UpdateUploadSettings(ctx context.Context, us api.UploadSettings) error {
+	data, err := json.Marshal(us)
+	if err != nil {
+		return fmt.Errorf("couldn't marshal the given value, error: %v", err)
+	}
+	return s.updateSetting(ctx, SettingUpload, string(data))
+}
+
+func (s *SQLStore) S3Settings(ctx context.Context) (ss api.S3Settings, _ error) {
+	value, err := s.fetchSetting(ctx, SettingS3)
+	if err != nil {
+		return api.S3Settings{}, err
+	}
+
+	if err := json.Unmarshal([]byte(value), &ss); err != nil {
+		s.logger.Panicf("failed to unmarshal s3 settings '%s': %v", value, err)
```

Comment:
A panic is a bit too harsh I think. Might be more graceful to log an error here and then return the defaults.